### ISSUE TITLE
The SAML assertion for "Shib-Identity-Provider" was null.

### ISSUE THREAD

**Issue by TanayKarve** [IQSS/dataverse#9328]
I am trying to integrate Shibboleth with dataverse.
I have the dataverse and shibboleth running as different services behind a reverse proxy.
I followed all steps in depth from the guide.
/Shibboleth.sso/Session has all the appropriate values.
I even looked into a similar issue on the google group, but my shibboleth2.xml does have AJP_ prefix set.
Even confirmed that attribute-map.xml is same as from the guide.

However, after logging in from my IDP, I am redirected to the blank account information page, and the logs say this:
`The SAML assertion for "Shib-Identity-Provider" was null. Please contact support.`

Am I missing something? How do I solve this?

**Comment by pdurbin** [comment]
@TanayKarve can you please post the output of /Shibboleth.sso/Session here? Or https://chat.dataverse.org

Or support@dataverse.org if you'd prefer to keep it private. Thanks!

**Comment by TanayKarve** [comment]
<img width="984" alt="image" src="https://user-images.githubusercontent.com/42843134/214926982-bae8960a-fce1-4771-899e-15d0d82b33d6.png">

**Comment by pdurbin** [comment]
@TanayKarve thanks! Here's what I see:

```
Miscellaneous
Session Expiration (barring inactivity): 479 minute(s)
Client Address: 10.137.173.242
SSO Protocol: urn:oasis:names:tc:SAML:2.0:protocol
Identity Provider: https://fed.huit.harvard.edu/idp/shibboleth
Authentication Time: 2023-01-26T14:39:34.564Z
Authentication Context Class: urn:oasis:names:tc:SAML:2.0:ac:classes:PasswordProtectedTransport
Authentication Context Decl: (none)

Attributes
affiliation: staff@harvard.edu;member@harvard.edu;employee@harvard.edu
displayName: Philip Durbin
eppn: 7af9aa0d0cbfc97e@harvard.edu
givenName: Philip
mail: philip_durbin@harvard.edu
sn: Durbin
```

Yours also has "Identity Provider" so I'm not sure why it isn't working. (I'm assuming you change the real value to "mysite.edu" but that's fine.)

It would be nice to see what this method prints out:

```
    public static void printAttributes(HttpServletRequest request) {
        List<String> shibValues = new ArrayList<>();
        if (request == null) {
            logger.fine("HttpServletRequest was null. No shib values to print.");
            return;
        }
        for (String attr : shibAttrs) {

            /**
             * @todo explain in Installers Guide that in order for these
             * attributes to be found attributePrefix="AJP_" must be added to
             * /etc/shibboleth/shibboleth2.xml like this:
             *
             * <ApplicationDefaults entityID="https://dataverse.org/shibboleth"
             * REMOTE_USER="eppn" attributePrefix="AJP_">
             *
             */
            Object attrObject = request.getAttribute(attr);
            if (attrObject != null) {
                shibValues.add(attr + ": " + attrObject.toString());
            }
        }
        logger.fine("shib values: " + shibValues);
    }
```

Can you please increase the logging level like this...

```
./asadmin set-log-levels edu.harvard.iq.dataverse.authorization.providers.shib.ShibUtil=FINE
```

... as explained at https://guides.dataverse.org/en/5.12.1/admin/troubleshooting.html#increasing-payara-logging

Then, please try again and look for "shib values" in server.log.

**Comment by TanayKarve** [comment]
Hmm, strange. I see this for shib values:
` shib values: []`

**Comment by pdurbin** [comment]
I know the comment says "change the entityID to your hostname"...

```
    <!-- FIXME: change the entityID to your hostname. -->
    <ApplicationDefaults entityID="https://dataverse.example.edu/sp"
        REMOTE_USER="eppn" attributePrefix="AJP_">
```

... but really it should be whatever your entityId is. From your earlier screenshot it seems like you have the older style that doesn't look like a URL.

MIT's is `urn:mace:incommon:mit.edu` for example, see https://incommon.org/custom/federation/info/entity.html?entityID=urn%3Amace%3Aincommon%3Amit.edu&technical=true

Assuming yours is the older, more distinguished `urn:mace...` form, can you please try it, like this...

```
    <ApplicationDefaults entityID="urn:mace:incommon:mit.edu"
        REMOTE_USER="eppn" attributePrefix="AJP_">
```

**Comment by TanayKarve** [comment]
@pdurbin So I changed entityID to urn:mace:incommon:mysite.edu, but my IDP organisation said that entityID needs to be dataplanet.mysite.edu/shibboleth-sp because urn:mace:incommon:mysite.edu is the id of the IDP.
They confirmed that SAML response is correctly being sent with all the values.
Do you think it might be an issue with my reverse proxy?  Here is my shib.conf:
```
# https://wiki.shibboleth.net/confluence/display/SHIB2/NativeSPApacheConfig

# RPM installations on platforms with a conf.d directory will
# result in this file being copied into that directory for you
# and preserved across upgrades.

# For non-RPM installs, you should copy the relevant contents of
# this file to a configuration location you control.
# don't pass paths used by Shibboleth to Payara
ProxyPassMatch ^/Shibboleth.sso !
ProxyPassMatch ^/shibboleth-ds !
# # pass everything else to Payara
#ProxyPass / ajp://localhost:8009/

#
# Load the Shibboleth module.
#
LoadModule mod_shib /usr/lib64/shibboleth/mod_shib_24.so

#
# Turn this on to support "require valid-user" rules from other
# mod_authn_* modules, and use "require shib-session" for anonymous
# session-based authorization in mod_shib.
#
ShibCompatValidUser On

#
# Ensures handler will be accessible.
#
<Location /Shibboleth.sso>
  AuthType None
  Require all granted
</Location>

#
# Used for example style sheet in error templates.
#
<IfModule mod_alias.c>
  <Location /shibboleth-sp>
    AuthType None
    Require all granted
  </Location>
  Alias /shibboleth-sp/main.css /usr/share/shibboleth/main.css
</IfModule>

#
# Configure the module for content.
#
# You MUST enable AuthType shibboleth for the module to process
# any requests, and there MUST be a require command as well. To
# enable Shibboleth but not specify any session/access requirements
# use "require shibboleth".
#
<Location /shib.xhtml>
  AuthType shibboleth
  ShibRequestSetting requireSession 1
  require valid-user
</Location>
# 

```

**Comment by pdurbin** [comment]
@TanayKarve woof. Honestly, as this point, I would try with https://samltest.id

It's been a little while since I tried but it should work. I described SAMLtest here: https://guides.dataverse.org/en/5.12.1/installation/shibboleth.html#exchange-metadata-with-your-identity-provider

The reason I added it to the guides is that sometimes people (like me) have trouble with their institutional identity provider and verifying that Dataverse works with SAMLtest helps people (like me) feel confident that they've configured Dataverse properly.

From a quick look your config looks fine but I may try to drag @donsizemore into this because he has much more current operational experience with Shib.

Anyway, please hang in there. We'll figure this out. At some point we can always reach out to the shib mailing list, which has been very helpful: https://shibboleth.net/mailman/listinfo/users

**Comment by donsizemore** [comment]
@TanayKarve Just curious, why aren't you passing traffic to Payara?

**Comment by TanayKarve** [comment]
@donsizemore okay, just uncommented `#ProxyPass / ajp://localhost:8009/`
but this is in shib.conf which is in a seperate container for shibboleth.
dataverse is in a seperate container which has payara running.
both services are under a reverse proxy using [traefik](https://doc.traefik.io/traefik/)
how do i forward from the shibboleth container to the dataverse container?

**Comment by pdurbin** [comment]
@TanayKarve since you're into modern things like containers, another thought for you: 😄 

- Set up OIDC on the Dataverse side: https://guides.dataverse.org/en/5.12.1/installation/oidc.html
- Set up Keycloak. (See #9234 but please note that it's just for dev and had a hard coded secret in it)
- Configure (in Dataverse) OIDC to talk to Keycloak.
- Configure Keycloak to talk to your institution's identity provider (using SAML)
- Profit!

No shibd anymore. Dataverse <-> OIDC <-> Keycloak <-> your IdP.

This is the direction we're going in the future anyway, probably. 😄 (See this [auth doc](https://docs.google.com/document/d/1mzPLbzAiED93hTXnTyw9rDBm7_dL1Qf4bPMg2j0c0NE/edit?usp=sharing)).

@poikilotherm has something similar at https://data.fz-juelich.de . OIDC but Unity instead of Keycloak.

**Comment by TanayKarve** [comment]
How do I see if Payara is receiving data?

**Comment by donsizemore** [comment]
There's an HTTP header that's not getting passed between Traefik and Shibboleth, but I forget what it is.

@4tikhonov solved this exact problem and described it to Phil and to me in the Community Slack, but it's the free version and I can't read back that far. Let me Google a bit.

**Comment by pdurbin** [comment]
> @4tikhonov solved this exact problem and described it to Phil and to me in the Community Slack, but it's the free version and I can't read back that far.

Gah! This is exactly why I opened this issue:

- #8712

As I said there, my vote is to shut down the community Slack and instead use https://chat.dataverse.org because it's [SLOPI](https://github.com/pdurbin/slopi-communication) (and free, and based on standards, etc.). I give everyone a little Matrix/Element badge at https://people.dataverse.org if they join. 😄 

The eventual goal is [SMACKI](http://blog.greptilian.com/2020/02/03/smacki/), some sort of searchable knowledge base, but we should start with transparent communication.

Anyway, let me go search my email. And I'll bug Slava next week to see if he can remember. Thanks, Don!

**Comment by pdurbin** [comment]
> Also, in the logs I see this:
> `| JkEnvVar: null|#]`

At https://groups.google.com/g/dataverse-community/c/avpF_dPQOPo/m/OYpE0tSsAAAJ @mbamouni from https://entrepot.recherche.data.gouv.fr (which also has a fancy containerized setup that @edzale and @tjouneau described in a recent community meeting) wrote:

"I solved my problem using AJP to make reverse proxy in httpd"

Does that help?

**Comment by pdurbin** [comment]
@TanayKarve in case you missed it, over at https://github.com/IQSS/dataverse-docker/issues/37#issuecomment-1407389716 @4tikhonov said:

"Hi @TanayKarve, you can find working integration of Dataverse and Shibboleth with traefik here: https://github.com/DataverseNO/dataverse-docker/tree/dataverse.no/distros/dataverse.no "

**Comment by pdurbin** [comment]
@TanayKarve did that help?

**Comment by TanayKarve** [comment]
Hey, sorry for the late response. Was working on my midterms haha.
I'm still trying to debug using the Dataverse.NO repo which has the shibboleth enabled composefile.
I am getting a new error now on my browser after I complete signing in with my SSO:
on /Shibboleth.sso/SAML2/POST ->
`opensaml::BindingException`
`SAML message delivered with POST to incorrect server URL.`

**Comment by Louis-wr** [comment]
If it helps I just uploaded the shibboleth configuration to 
https://github.com/DataverseNO/dataverse-docker/tree/dataverse.no/shibboleth

**Comment by pdurbin** [comment]
@TanayKarve now it's my turn to apologize. I didn't see this until just now. I'm afraid that I haven't seen that error. Again, I don't have experience using Shibboleth in Docker.

However... @Louis-wr does! (Thanks for pushing that config!) Perhaps the two of you could get on a call? I'm happy to join.

**Comment by TanayKarve** [comment]
@Louis-wr , have you seen this error? (After signing in with SSO)

**opensaml::BindingException**


```opensaml::BindingException at (http://dataplanet.mysite.edu/Shibboleth.sso/SAML2/POST)```
```SAML message delivered with POST to incorrect server URL.```

**Comment by pdurbin** [comment]
@TanayKarve it sounds like @4tikhonov can meet as soon as tomorrow. @Louis-wr is more than welcome, of course! Let's try to get a Zoom going.

On a related note, I don't know when this would be prioritized but I only run Docker on my laptop, never on a server so as we discussed I opened this issue to add a "Docker mode" to our EC2 spin up script:

- https://github.com/GlobalDataverseCommunityConsortium/dataverse-ansible/issues/268
